### DISCUSSION TITLE
Upgrading to 9.22.4 aws s3 files say "You don't have permission to access this."

### DISCUSSION THREAD

**Discussion by elplaza** [directus/directus#17286] [Archive]
### Describe the Bug

My files are on AWS S3 and my configuration is like this:
```sh
STORAGE_LOCATIONS	aws
STORAGE_AWS_DRIVER	s3
STORAGE_AWS_KEY	        <my-aws-key>
STORAGE_AWS_SECRET	<my-aws-secret>
STORAGE_AWS_BUCKET	<my-aws-bucket>
STORAGE_AWS_REGION	<my-aws-region>
```

All was working fine with docker image 9.21.0.

Upgrading to 9.22.4 all files are "Forbidden" both for read and write operations.
UI network calls like:
`GET <directus>/assets/94f296a1-fcc7-468a-ba6b-da86dadc6395`
have 403 response with:
` {"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}`

### To Reproduce

Start with docker image 9.21.0 with storage configuration like this:
```sh
STORAGE_LOCATIONS	aws
STORAGE_AWS_DRIVER	s3
STORAGE_AWS_KEY	        <my-aws-key>
STORAGE_AWS_SECRET	<my-aws-secret>
STORAGE_AWS_BUCKET	<my-aws-bucket>
STORAGE_AWS_REGION	<my-aws-region>
```

Upgrade docker image to 9.22.4

### Errors Shown

UI network calls like:
`GET <directus>/assets/94f296a1-fcc7-468a-ba6b-da86dadc6395`
have 403 response with:
` {"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}`

### What version of Directus are you using?

9.22.4

### What version of Node.js are you using?

v18.13.0

### What database are you using?

MySQL

### What browser are you using?

Chrome

### How are you deploying Directus?

Docker

**Comment by paescuj** [comment]
It seems like you're missing the `STORAGE_<LOCATION>_ENDPOINT` configuration.
Please consult the [documentation](https://docs.directus.io/self-hosted/config-options.html#file-storage).
Also this issue might be helpful: #16876

Closing for now as it seems to be the result of a misconfiguration.

**Comment by RegisWeduLabs** [comment]
Hi - I have the configuration on my side - but I am getting the exact same error since upgrading to 9.22.4.  Adding https:// in front of the STORAGE_<LOCATION>_ENDPOINT fixed the issue for me. (https://github.com/directus/directus/issues/16876#issuecomment-1362297055)

**Comment by elplaza** [comment]
I added the STORAGE_AWS_ENDPOINT environment variable but I get always the same 403 error.

My configuration now is:
```sh
STORAGE_LOCATIONS	   aws
STORAGE_AWS_DRIVER	   s3
STORAGE_AWS_KEY	          <my-aws-key>
STORAGE_AWS_SECRET	  <my-aws-secret>
STORAGE_AWS_BUCKET	  <my-aws-bucket>/<my-aws-basedir>
STORAGE_AWS_REGION	  <my-aws-region>
STORAGE_AWS_ENDPOINT  <my-aws-endpoint>
```

I tried with all the following endpoint without luck:
```sh
- s3.amazonaws.com
- https://s3.amazonaws.com
- https://s3.<my-aws-region>.amazonaws.com
- <my-aws-bucket>.s3.<my-aws-region>.amazonaws.com
- https://<my-aws-bucket>.s3.<my-aws-region>.amazonaws.com
- https://<my-aws-bucket>.s3.<my-aws-region>.amazonaws.com/
```
This env var was not required, now it is? Why?
It would be great if you tell me where can i find this endpoint on aws.
Sorry but reading docs it is not so clear what endpoint are you talking about . . .

**Reply by paescuj** [reply]
Transformed the issue into a discussion, so others are able to help as well.
I'm afraid I cannot be of further assistance here, as I've too little experience with AWS storage myself.

**Reply by elplaza** [reply]
@paescuj can be the `<my-aws-basedir>` the problem?
In AWS S3 we have the bucket `<my-aws-bucket>`, the object `<my-file-content>` and the key `<my-aws-key>`. 
If we want to organize files in a "directories tree like", for ex. `avengers/ironman/mark1.jpg`, we have that all the tree is the key: `<my-aws-key> = avengers/ironman/mark1.jpg`.
My case is similar to this: I have a `<my-aws-basedir>` as "file prefix" so the resulting aws object key must be `<my-aws-basedir>/<filename_disk>`.

In Directus v9.21.0 this "directory" was in the STORAGE_AWS_BUCKET env variable, maybe it's not the same in Directus v9.22.4?

**Reply by elplaza** [reply]
My use case is almost basic also to reproduce:
1) Directus v9.22.4
2) storage env vars to store files on AWS S3 bucket (ex. bucket name: `superheroes`)
3) create a collection on Directus with field `file`
4) uploading a file (ex. filename_disk = `00022895-34c7-4591-a007-7db98e505085.jpg`) this should be stored on AWS S3 bucket `superheroes` with key = `avengers/00022895-34c7-4591-a007-7db98e505085.jpg`
5) the file you just uploaded should be visible on Directus UI

**Reply by elplaza** [reply]
@rijkvanzanten you made the last commit on [this file](https://github.com/directus/directus/blob/main/packages/storage-driver-s3/src/index.ts) . . . can you help me? Can the `STORAGE_<LOCATION>_FORCE_PATH_STYLE` env var help me?

**Reply by paescuj** [reply]
> can be the `<my-aws-basedir>` the problem?

Yes, I don't think that you can specify this inside the `STORAGE_AWS_BUCKET` config.
To specify a file prefix, I think you're supposed to use `STORAGE_AWS_ROOT` instead.

Please also see the official documentation:
- `BUCKET`:  https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/getobjectcommandinput.html#bucket
- `FORCE_PATH_STYLE`: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/s3clientconfig.html#forcepathstyle

**Reply by elplaza** [reply]
@paescuj thanks for the reply :hugs:
Reading your answer and this [article about](https://medium.com/@e.osama85/the-difference-between-the-amazon-s3-path-style-urls-and-virtual-hosted-style-urls-4fafd5eca4db) I made following attemps setting `STORAGE_AWS_FORCE_PATH_STYLE = true` and `STORAGE_AWS_ENDPOINT` to:
- Global S3 endpoint: `https://s3.amazonaws.com/<my-aws-bucket>`
- Regional S3 endpoint: `https://s3-<my-aws-region>.amazonaws.com/<my-aws-bucket>`

both with and without the trailing slash.

I tried also to add the STORAGE_AWS_ROOT env var but without luck.

@paescuj @rijkvanzanten can some of you give me a working configuration to obtain [this use case](https://github.com/directus/directus/discussions/17286#discussioncomment-4775026)?
I simply want that the files uploaded by Directus UI are loaded on AWS S3 bucket `superheroes` with object key `avengers/00022895-34c7-4591-a007-7db98e505085.jpg`.

**Reply by rijkvanzanten** [reply]
What Pascal mentioned above is correct. The `<my-aws-basedir>` isn't part of the `BUCKET` configuration, but should instead be configured in `ROOT`.

**Reply by paescuj** [reply]
Thanks @rijkvanzanten for your confirmation! I guess in the end it should look something like this:
```bash
STORAGE_LOCATIONS="aws"
STORAGE_AWS_DRIVER="s3"

STORAGE_AWS_KEY="<my-aws-key>"
STORAGE_AWS_SECRET="<my-aws-secret>"

# Not quite sure about the endpoint address...
STORAGE_AWS_ENDPOINT="https://s3.amazonaws.com"
STORAGE_AWS_REGION="<my-aws-region>"
STORAGE_AWS_BUCKET="superheroes"

STORAGE_AWS_ROOT="avengers"
```

**Reply by elplaza** [reply]
Good news: I found the working configuration for my use case
Bad news: I made attempts after attempts until it worked; I don't understand **why** it worked

However the working configuration for my use case is:
```sh
STORAGE_LOCATIONS="aws"
STORAGE_AWS_DRIVER="s3"
STORAGE_AWS_KEY="<my-aws-key>"
STORAGE_AWS_SECRET="<my-aws-secret>"
STORAGE_AWS_REGION="<my-aws-region>"
STORAGE_AWS_BUCKET="superheroes"
STORAGE_AWS_ENDPOINT="https://s3-<my-aws-region>.amazonaws.com"
STORAGE_AWS_ROOT="avengers"
STORAGE_AWS_FORCE_PATH_STYLE="true"
```

@paescuj @rijkvanzanten thanks for your precious help!

**Reply by paescuj** [reply]
@elplaza I'm glad to hear that it's working for you now!
Unfortunately, configuring S3 with special requirements can be quite tricky...

A resource which might help in such cases is the official documentation of AWS's SDK (which is what's used by Directus): https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/s3clientconfig.html

**Comment by GioKaaja** [comment]
@paescuj it will be super helpful from your side if you could help elplaza solving the mistery!!

**Comment by jcam** [comment]
For my configuration, I had previously set the endpoint to s3.amazonaws.com, which _matches the documentation_.
This worked fine for 9.18.1  through ~ 9.21, but was recently reported broken after the 9.22-ish upgrade (i found that I had left one environment set to automatically update itself)

<img width="732" alt="image" src="https://user-images.githubusercontent.com/98744/215910131-9cc159fd-8574-480e-96a1-3d05f70c10e8.png">

Changing it to https:// did resolve it, but this is a breaking change!

**Reply by jcam** [reply]
https://github.com/directus/directus/blob/d928275447d33351af8d0af246bbc68d90783cb8/packages/storage-driver-s3/src/index.ts#L56

IMO this should default to https, rather than defaulting to http.

**Reply by paescuj** [reply]
@rijkvanzanten Unfortunately, I don't know enough about why and what has been changed for the S3 config in the context of #16825. Do you think that the documentation or even some default values would have to be adapted in this respect?

**Reply by jcam** [reply]
@paescuj let me know if I should open a separate issue and/or a PR to change the default to https to match previous releases (draft PR here: https://github.com/directus/directus/pull/17372)

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten Unfortunately, I don't know enough about why and what has been changed for the S3 config in the context of #16825. Do you think that the documentation or even some default values would have to be adapted in this respect?

We upgraded from `aws-sdk` (v2) to `@aws-sdk/client-s3` to (v3) which is where most of these config default changes stem from. Can we confirm that aws-sdk had `https` as its default?

**Reply by paescuj** [reply]
> We upgraded from aws-sdk (v2) to @aws-sdk/client-s3 to (v3) which is where most of these config default changes stem from.

All right, thanks for the explanation!

> Can we confirm that aws-sdk had https as its default?

Did a quick test and apparently `https` indeed used to be the default. This is what I got with the default values (see also [`AWS.Endpoint`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Endpoint.html)):
```
endpoint: Endpoint {
  protocol: 'https:',
  host: 's3.amazonaws.com',
  port: 443,
  hostname: 's3.amazonaws.com',
  pathname: '/',
  path: '/',
  href: 'https://s3.amazonaws.com/'
}
```

I think we should be able to switch back to that behavior with something like #17372, while leaving a notice in the release notes.
Most users have probably specified the endpoint including protocol anyway.

**Comment by jgiron09** [comment]
I had the same issue, but was solved adding "read" permission to **direcus_files** system collection
<img width="874" alt="image" src="https://github.com/user-attachments/assets/1f7b1334-b9f9-4230-b0b7-8bfcb1e0bfed">
